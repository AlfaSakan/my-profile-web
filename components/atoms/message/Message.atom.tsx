import ArrowLeftIcon from '@material-ui/icons/ArrowLeft';
import ArrowRightIcon from '@material-ui/icons/ArrowRight';
import DoneAllIcon from '@material-ui/icons/DoneAll';
import { User } from 'models/user/user.model';
import React from 'react';
import { convertNumberToTime } from 'utils/converter.util';

interface IProps {
  message: string;
  time: number;
  isIcon?: boolean;
  isLeft?: boolean;
  sender?: User;
  senderColor?: string;
}

const Message: React.FC<IProps> = ({
  message,
  time,
  isIcon = true,
  isLeft = true,
  sender,
  senderColor,
}) => {
  return (
    <div className="flex-col bg-white h-fit min-w-fitCostume max-w-xs sm:max-w-xl px-2 py-1 relative rounded-md text-sm">
      {isIcon &&
        (isLeft ? (
          <div className="absolute -top-2 -left-[0.9rem]">
            <ArrowLeftIcon style={{ fill: 'white', fontSize: 30 }} />
          </div>
        ) : (
          <div className="absolute -top-2 -right-[0.9rem]">
            <ArrowRightIcon style={{ fill: 'white', fontSize: 30 }} />
          </div>
        ))}
      {sender && isLeft && (
        <p className="text-xs font-semibold" style={{ color: senderColor }}>
          {sender?.name}
        </p>
      )}
      <p>{message}</p>
      <div className="items-center gap-1 self-end">
        <p className="text-xs">{convertNumberToTime(time)}</p>
        <DoneAllIcon style={{ fontSize: 14 }} />
      </div>
    </div>
  );
};

export default Message;
