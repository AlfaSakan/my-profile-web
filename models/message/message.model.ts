export interface Message {
  message_id: number;
  chat_room_id: number;
  sender_id: number;
  status: string;
  type: string;
  message: string;
  created_at: number;
}
