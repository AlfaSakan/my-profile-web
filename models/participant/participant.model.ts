import { User } from 'models/user/user.model';

export interface Participant extends User {
  isFriend?: boolean;
}
