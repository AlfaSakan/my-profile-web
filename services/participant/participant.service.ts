import { PARTICIPANT_ROUTE } from 'constants/route.constant';
import { Response } from 'models/response/response.model';
import { User } from 'models/user/user.model';
import fetchBase from 'services/api/api.service';

const participantService = () => {
  const { query } = fetchBase();

  const getListParticipant = async (chatRoomId: string) => {
    try {
      const res = (await query(
        `${PARTICIPANT_ROUTE}/${chatRoomId}`,
        'GET'
      )) as Response<User[]>;

      if (res.status !== 200 || res.data === null) {
        return;
      }

      return res;
    } catch (error) {
      console.log('useParticipantApi getListParticipant error', error);
    }
  };

  return {
    getListParticipant,
  };
};

export default participantService;
