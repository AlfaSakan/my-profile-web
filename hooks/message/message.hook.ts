import { useState } from 'react';
import { messageService } from 'services';
import { MessageParams } from 'services/message/message.service';

const useMessageHook = () => {
  const [message, setMessage] = useState('');

  const { postMessage } = messageService();

  const messageHandler = (value: string) => {
    setMessage(value);
  };

  const onSubmitMessage =
    (sender_id: string, chat_room_id: string, type?: string) => async () => {
      try {
        const messageRequest: MessageParams = {
          message,
          chat_room_id,
          sender_id,
          type,
        };

        const res = await postMessage(messageRequest);

        setMessage('');
        return res;
      } catch (error) {
        console.log('onSubmitMessage error', error);
      }
    };

  return {
    message,
    messageHandler,
    onSubmitMessage,
  };
};

export default useMessageHook;
