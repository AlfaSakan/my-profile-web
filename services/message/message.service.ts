import { MESSAGE_ROUTE } from 'constants/route.constant';
import { Message } from 'models/message/message.model';
import { Response } from 'models/response/response.model';
import fetchBase from 'services/api/api.service';

export interface MessageParams {
  chat_room_id: string;
  sender_id: string;
  type?: string;
  message: string;
}

const messageService = () => {
  const { query, mutation } = fetchBase();

  const getMessages = async (chatRoomId: number) => {
    try {
      const res = (await query(
        `${MESSAGE_ROUTE}/${chatRoomId}`,
        'GET'
      )) as Response<Message[]>;

      return res;
    } catch (error) {
      console.log('useMessageApi getMessages error', error);
    }
  };

  const postMessage = async (request: MessageParams) => {
    try {
      const res = (await mutation(
        MESSAGE_ROUTE,
        'POST',
        request
      )) as Response<Message>;

      if (res.status !== 201) return;

      return res;
    } catch (error) {
      console.log('useMessageApi postMessage error', error);
    }
  };

  return {
    getMessages,
    postMessage,
  };
};

export default messageService;
