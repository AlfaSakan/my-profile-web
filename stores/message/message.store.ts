import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { MESSAGE_KEY } from 'constants/storage.constant';
import { Message } from 'models/message/message.model';

export interface MessageStore extends Message {}

export const messageStoreInitialState: MessageStore[] = [];

const messageSlice = createSlice({
  name: 'message store',
  initialState: messageStoreInitialState,
  reducers: {
    retrieveMessages: (_, action: PayloadAction<MessageStore[]>) => {
      return [...action.payload];
    },

    insertMessage: (state, action: PayloadAction<MessageStore>) => {
      const messages = [...state, action.payload];

      localStorage.setItem(MESSAGE_KEY, JSON.stringify(messages));

      return messages;
    },

    deleteMessageStore: () => {
      localStorage.removeItem(MESSAGE_KEY);

      return [];
    },
  },
});

export const { insertMessage, retrieveMessages, deleteMessageStore } =
  messageSlice.actions;

export default messageSlice.reducer;
