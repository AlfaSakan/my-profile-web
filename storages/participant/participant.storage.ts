import { PARTICIPANT_KEY } from 'constants/storage.constant';
import { Participant } from 'models/participant/participant.model';

const insertListParticipants = (participants: Participant[]) => {
  localStorage.setItem(PARTICIPANT_KEY, JSON.stringify(participants));
};

const insertParticipant = (participant: Participant) => {
  const participants = retrieveParticipants();

  let isExisted = false;

  for (const part of participants) {
    isExisted = part.user_id === participant.user_id;
  }

  if (isExisted) return;

  insertListParticipants(participants);
};

const retrieveParticipants = () => {
  const participants = localStorage.getItem(PARTICIPANT_KEY);

  if (!participants) return {} as Participant[];

  return JSON.parse(participants) as Participant[];
};

const removeParticipants = () => localStorage.removeItem(PARTICIPANT_KEY);

const addFriend = (participant: Participant) => {
  const participants = retrieveParticipants();

  for (const part of participants) {
    if (part.user_id === participant.user_id) {
      part.isFriend = true;
    }
  }

  insertListParticipants(participants);
};

const participantStorage = {
  retrieveParticipants,
  insertListParticipants,
  insertParticipant,
  removeParticipants,
  addFriend,
};

export default participantStorage;
