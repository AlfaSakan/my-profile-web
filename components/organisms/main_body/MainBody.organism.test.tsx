import { Message } from 'models/message/message.model';
import { Participant } from 'models/participant/participant.model';
import renderer from 'react-test-renderer';
import MainBody from './MainBody.organism';

describe('HeaderBody component test', () => {
  const messages: Message[] = [
    {
      chat_room_id: '1',
      created_at: 123456789,
      message: '123456789',
      message_id: '1',
      sender_id: '1',
      status: 'chat',
      type: '',
    },
  ];

  const userData: Participant[] = [
    {
      country_code: '+62',
      user_id: '1',
      name: 'admin',
      created_at: 123456789,
      image_url: '',
      phone_number: '8123456789',
      status: '',
      updated_at: 1213456789,
    },
  ];

  it('should render correctly', () => {
    const tree = renderer
      .create(<MainBody userId="1" messages={messages} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const tree = renderer
      .create(<MainBody userId="1" messages={messages} userData={userData} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
