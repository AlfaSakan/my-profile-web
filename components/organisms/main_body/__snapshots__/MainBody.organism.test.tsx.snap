// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBody component test should render correctly 1`] = `
<div
  className="flex flex-col-reverse flex-1 sm:px-20 px-5 gap-1 pb-5 overflow-y-scroll"
>
  <div
    className="self-end"
  >
    <div
      className="flex-col bg-white h-fit min-w-fitCostume max-w-xs sm:max-w-xl px-2 py-1 relative rounded-md text-sm"
    >
      <div
        className="absolute -top-2 -right-[0.9rem]"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          style={
            Object {
              "fill": "white",
              "fontSize": 30,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M10 17l5-5-5-5v10z"
          />
        </svg>
      </div>
      <p
        className="text-sm"
      >
        123456789
      </p>
      <div
        className="items-center gap-1 self-end"
      >
        <p
          className="text-xs undefined"
        >
          17:17
        </p>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          style={
            Object {
              "fontSize": 14,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M18 7l-1.41-1.41-6.34 6.34 1.41 1.41L18 7zm4.24-1.41L11.66 16.17 7.48 12l-1.41 1.41L11.66 19l12-12-1.42-1.41zM.41 13.41L6 19l1.41-1.41L1.83 12 .41 13.41z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`HeaderBody component test should render correctly 2`] = `
<div
  className="flex flex-col-reverse flex-1 sm:px-20 px-5 gap-1 pb-5 overflow-y-scroll"
>
  <div
    className="self-end"
  >
    <div
      className="flex-col bg-white h-fit min-w-fitCostume max-w-xs sm:max-w-xl px-2 py-1 relative rounded-md text-sm"
    >
      <div
        className="absolute -top-2 -right-[0.9rem]"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          style={
            Object {
              "fill": "white",
              "fontSize": 30,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M10 17l5-5-5-5v10z"
          />
        </svg>
      </div>
      <p
        className="text-sm"
      >
        123456789
      </p>
      <div
        className="items-center gap-1 self-end"
      >
        <p
          className="text-xs undefined"
        >
          17:17
        </p>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          style={
            Object {
              "fontSize": 14,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M18 7l-1.41-1.41-6.34 6.34 1.41 1.41L18 7zm4.24-1.41L11.66 16.17 7.48 12l-1.41 1.41L11.66 19l12-12-1.42-1.41zM.41 13.41L6 19l1.41-1.41L1.83 12 .41 13.41z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
